src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -19,7 +23,54 @@ internal static class DeserializationHelpers
         internal static void ApplyProperty(ODataProperty property, IEdmStructuredTypeReference resourceType, object resource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
         {
-            IEdmProperty edmProperty = resourceType.FindProperty(property.Name);
+            WebApiRequestMessage request = null;
+            if(readContext != null)
+            {
+                request = (WebApiRequestMessage)readContext.InternalRequest;
+            }
+
+            bool enableCaseInsensitiveModelBinding = false;
+
+            if(request != null)
+            {
+#if NETCORE
+                // In AspNetCore, we add our setting by adding a service to the ServiceCollection as below:
+                // services.TryAddEnumerable(ServiceDescriptor.Singleton<IODataModelBinderSettings>(new ODataModelBinderSettings { EnableCaseInsensitiveModelBinding = true }));
+                // We need to resolve the service so that we can get our setting.
+                var settings = request.innerRequest.HttpContext.RequestServices.GetService<IODataModelBinderSettings>();
+                if(settings != null)
+                {
+                    enableCaseInsensitiveModelBinding = settings.EnableCaseInsensitiveModelBinding;
+                }
+#else
+                // In AspNet classic, we add our setting to HttpConfiguration.Properties as below:
+                // configuration.Properties[EnableCaseInsensitiveModelBindingKey] = true;
+                var container = request.innerRequest.GetConfiguration();
+                if(container != null)
+                {
+                    enableCaseInsensitiveModelBinding = container.HasEnabledCaseInsensitiveModelBinding();
+                }
+#endif
+            }
+
+            IEdmProperty edmProperty;
+
+            if (enableCaseInsensitiveModelBinding)
+            {
+                IDictionary<string, string> propertyNamesMapping = new Dictionary<string, string>();
+                var sProperties = resourceType.StructuralProperties();
+                foreach (var prop in sProperties)
+                {
+                    propertyNamesMapping.Add(prop.Name.ToUpperInvariant(), prop.Name);
+                }
+                string name;
+                var findPropertyName = propertyNamesMapping.TryGetValue(property.Name.ToUpperInvariant(), out name) ? name : property.Name;
```

Comment:
Alternatively, you could use a `HashSet` initialized with `StringComparer.OrdinalIgnoreCase` instead of a full-blown `Dictionary`

Reply:
I like the use of the hashset here.

---
In reply to: [487683925](https://github.com/OData/WebApi/pull/2249#discussion_r487683925) [](ancestors = 487683925)

Reply:
Unfortunately `Hashset<string>.TryGetValue` is not available in `netstandard 2.0` https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.trygetvalue?view=netcore-3.1

Reply:
Technically you could still use `hashset.Contains(property.Name)` to achieve the same result - which I think is still better than using a dict

Reply:
`hashset.Contains(property.Name)` returns a `bool`. However we need to extract the value from the hashset. This can only be achieved by iteration O(n). While it is a solution to the problem, it kind of defeats the purpose of using a hashset in the first place.

Reply:
Oh yeah you're right. Had forgotten that part

Looking at this again though, what's the benefit of creating the Dict here? Since you are already looping anyway, you might as well get rid of the Dict altogether and just do the check inside the loop instead of populating a Dict in the loop THEN doing the check. That or consider Mike's comment on creating the mapping earlier and reusing it if there are multiple props to check for

Reply:
#Resolved